flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -202,23 +204,122 @@ class BoxValue : public AbstractBox, public AbstractArrayBox {
   }
 
   BoxValue cloneElement(mlir::Value newBase) const {
-     return {newBase, len, params, {}, {}};
+    return {newBase, len, params, {}, {}};
   }
 
   mlir::Value getLen() const { return len; }
 
-  const llvm::SmallVectorImpl<mlir::Value> &getLenTypeParams() const {
-    return params;
-  }
+  llvm::ArrayRef<mlir::Value> getLenTypeParams() const { return params; }
 
   friend llvm::raw_ostream &operator<<(llvm::raw_ostream &, const BoxValue &);
   LLVM_DUMP_METHOD void dump() const { llvm::errs() << *this; }
 
 protected:
-  mlir::Value len; // box is CHARACTER
+  mlir::Value len;                          // box is CHARACTER
   llvm::SmallVector<mlir::Value, 2> params; // LENs, box is derived type
 };
 
+/// Base class for values associated to a fir.box or fir.ref<fir.box>.
+class AbstractIrBox : public AbstractBox {
+public:
+  AbstractIrBox(mlir::Value addr) : AbstractBox{addr} {}
+  /// Get the fir.box<type> part of the address type.
+  fir::BoxType getBoxTy() const {
+    auto type = getAddr().getType();
+    if (auto pointedTy = fir::dyn_cast_ptrEleTy(type))
```

Comment:
This is removing the outer reference if there is one. The rational is that the descriptor associated to a `MutableBoxValue` is always in memory (`fir.ref<fir.box<T>>`) because it can change, but `IrBoxValue` descriptor cannot change, so they are tracked by value (`fir.box<T>`). So this function, that returns `fir.box<T>` type, is just smoothing the out this difference for the rest of `AbstractIrBox` helper functions that do not care about it, and just want to reason about the box type.

Reply:
OK. So I'm not sure why we need both a `BoxValue` and an `IrBoxValue` as they apparently deal with the same sort of variables and their attributes.

Reply:
That's a good point, I think `BoxValue` needs to be retired.
They look the same, but `BoxValue` has some properties that makes it unusable for some entities that `IrBoxValue` support (such as optional assumed shapes):
- If it is an array, it expects that the extents are pre-read in the extents vector (it inherits the AbstractArrayBoxValue that has this property).
- Part of the lowering code, and I may be to blame a bit here, expects that `BoxValue` would be used for parametrized derived types (e.g., it is called `Derived` in the symbol box).
I propose just getting rid of it and then renaming `IrBoxValue` to `BoxValue`.